File: src/algorithmic_toolbox/week4/Majority.java
package algorithmic_toolbox.week4;

import java.util.Scanner;

public class Majority {

  private static int findMajority(int[] arr, int lo, int hi) {
    if (hi <= lo) {
      return arr[lo];
    }

    int mid = lo + (hi - lo) / 2;
    int maj1 = findMajority(arr, lo, mid);
    int maj2 = findMajority(arr, mid + 1, hi);

    return merge(arr, maj1, maj2, lo, hi);
  }

  private static int merge(int[] arr, int m1, int m2, int lo, int hi) {
    int m1Count = 0, m2Count = 0;
    int majCount = (hi - lo + 1) / 2 + 1;

    for (int i = lo; i <= hi; i++) {
      if      (arr[i] == m1) { m1Count++; }
      else if (arr[i] == m2) { m2Count++; }
    }

    if (m1Count >= majCount) {
      return m1;
    } else if (m2Count >= majCount) {
      return m2;
    } else {
      return -1;
    }
  }
  public static void main(String[] args) {
    Scanner scanner = new Scanner(System.in);
    int size = scanner.nextInt();

    int[] nums = new int[size];
    for (int i = 0; i < size; i++) {
      nums[i] = scanner.nextInt();
    }
    scanner.close();

    int maj = findMajority(nums, 0, nums.length - 1);
    System.out.println((maj == -1) ? 0 : 1);
  }
  
}

File: src/algorithmic_toolbox/week4/majority.py
from typing import List


def find_majority(lst : List[int], lo: int, hi: int) -> int:
    if (hi <= lo):
        return lst[lo]
    
    mid = lo + (hi - lo) // 2
    maj1 = find_majority(lst, lo, mid)
    maj2 = find_majority(lst, mid + 1, hi)
    return _merge(lst, maj1, maj2, lo, hi)


def _merge(lst: List[int], m1: int, m2: int, lo: int, hi: int) -> int:
    m1_count = m2_count = 0
    maj_count = (hi - lo + 1) // 2 + 1

    for i in range(lo, hi + 1):
        if lst[i] == m1:
            m1_count += 1
        elif lst[i] == m2:
            m2_count += 1
    
    if m1_count >= maj_count:
        return m1
    elif m2_count >= maj_count:
        return m2
    else:
        return -1


if __name__ == '__main__':
    input()  # size not required
    nums: List[int] = list(map(int, input().split()))
    maj = find_majority(nums, 0, len(nums) - 1)

    print(0 if maj <= -1 else 1)
